tests/network/port_forward.go

```diff
@@ -0,0 +1,139 @@
+/*
+ * This file is part of the kubevirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"fmt"
+	"os/exec"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	k8sv1 "k8s.io/api/core/v1"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/tests"
+	"kubevirt.io/kubevirt/tests/console"
+	"kubevirt.io/kubevirt/tests/libvmi"
+)
+
+var _ = SIGDescribe("[Serial] Port-forward", func() {
+	var (
+		err        error
+		virtClient kubecli.KubevirtClient
+	)
+
+	BeforeEach(func() {
+		virtClient, err = kubecli.GetKubevirtClient()
+		tests.PanicOnError(err)
+
+		tests.BeforeTestCleanup()
+	})
+
+	Context("VMI With masquerade binding", func() {
+		const localPort = 1500
+		var (
+			portForwardCmd *exec.Cmd
+			vmi            *v1.VirtualMachineInstance
+			VMIport        int
+			ports          []v1.Port
+		)
+
+		JustBeforeEach(func() {
+			vmi = createCirrosVMIWithPortsAndBlockUntilReady(virtClient, ports)
+			tests.StartHTTPServer(vmi, VMIport)
+
+			vmiPod := tests.GetRunningPodByVirtualMachineInstance(vmi, tests.NamespaceTestDefault)
+			Expect(vmiPod).ToNot(BeNil())
+			portForwardCmd, err = portForwardCommand(vmiPod, localPort, VMIport)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(portForwardCmd.Start()).To(Succeed())
+		})
+
+		When("performing port-forward from a local port to a VMI's pod declared port", func() {
+			const declaredPort = 1501
+			BeforeEach(func() {
+				ports = append(ports, v1.Port{Port: declaredPort})
+				VMIport = declaredPort
+			})
+
+			It("should reach the vmi", func() {
+				defer func() {
+					Expect(killPortForwardCommand(portForwardCmd)).To(Succeed())
+				}()
+
+				By(fmt.Sprintf("checking that service running on port %d can be reached", declaredPort))
+				Eventually(func() error {
```

Comment:
I think the VM can be considered ready (by `createCirrosVMIWithPortsAndBlockUntilReady()`) even though it is still haven't finish It's booting process. The VM is considered ready if there was a successfull login (see [here](https://github.com/kubevirt/kubevirt/blob/master/tests/utils.go#L3064)). Login to the VMI can be done even when the VMI is still booting up,  So the HTTP server might not have started when we reach the curl command. So I thought to use eventually to wait for that. 
In my local env the tests usually won't pass without `Eventually` and constantly passing with `Eventually`  and I think it is because of that.

Reply:
You start the http server in the `BeforeEach` so I don't understand how the server can still be down in the stage of the `curl`.
We use `WaitUntilVMIReady` in most of our tests and we don't need eventually there. For example - [this](https://github.com/kubevirt/kubevirt/blob/a532f588a293d0aee91c22161189214f2b6c1850/tests/network/vmi_networking.go#L767).
I believe the reason you need eventually is different. Please debug it.

Reply:
You are right. The reason I gave does not holds because we couldn't start the http server without the vmi to be ready and finish boot. 
It could be also that the port-forward command takes time to work. But Since I can not wait for it to finish and return a OK status, and I have to wait somewhere i think waiting in the Eventually won't be so bad. 

Even In the example you shared there are actually a lot of timeouts as part of the `verifyClientServerConnectivity()` func and the fact that it is using expecters which uses timeouts.

The only suspicious thing I thought of was that the port-forward command from a previous test still ha havn't finished (like you noted [here](https://github.com/kubevirt/kubevirt/pull/5431/commits/de977707f5a5071458ed10a0e7a747cd116a0da6#r636101749) but this is not true as without eventually it fails also when I run only one test.

What do you think?

Reply:
> It could be also that the port-forward command takes time to work
Let's not depend on assumption. I think we should understand why the first tries are not working.

> Even In the example you shared there are actually a lot of timeouts
There is a difference between a timeout and an eventually.
Timeout means the operation may take time, but failures are not accepted.
`Eventually` means the operation may fail several time before it succeeds.

Using `Eventually` may hide bugs. IMO, it's ok to use it only if we understand the exact reason why it is needed.

Reply:
Understand.
Thanks for explaining.

still the command takes sometime to act, and in order for the test to pass we need to wait a bit for it, before we try the curl.
I removed the `Eventually` and added 3 seconds sleep before the curl. it should be enough. 
that way we will still wait but not ignore failures.
What do you think?

Reply:
Removed the `Eventually` and now waiting for the port-forward cmd output.
https://github.com/kubevirt/kubevirt/compare/699f4e38bbaa7be7bf79da628385590f98f1faa4..2d512ffdf7688e2cd00402f51d783d38e9ce7913